lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/ts_node_lowering.cpp

```diff
@@ -319,9 +319,17 @@ class TSNodeLowering : public TSNodeLoweringInterface {
             ? LowerBuiltin(at::aten::cudnn_convolution_transpose_backward,
                            arguments)
             : LowerBuiltin(at::aten::cudnn_convolution_backward, arguments);
-    // cudnn_convolution_backward/cudnn_convolution_transpose_backward only
-    // returns 2 tensors
-    result.push_back(nullptr);
+
+    // N.B. rank is specialized even with dynamic shapes
+    // all axes but channel to reduce to the bias size
+    auto grad_rank = dynamic_cast<const ir::TsNode*>(operands[0].node)->shape(operands[0].index).dim();
+    std::vector<int64_t> axes(grad_rank);
+    std::iota(axes.begin(), axes.end(), 0);
+    TORCH_INTERNAL_ASSERT(grad_rank >= 3, "Convolution outputs should have 3+ dimensions");
+    axes.erase(axes.begin() + 1);
+    auto axes_val = loctx()->graph()->insertConstant(axes);
+    auto bias_grad = loctx()->graph()->insert(at::aten::sum, {loctx()->GetOutputOp(operands[0]), axes_val});
```

Comment:
I lack the context for why this change is necessary- we're essentially saying that we need to sum away all but the first dimension of the grad, which i guess is channels?  why isn't this isn't done for us by autograd?

Reply:
We may just get the last output from the generic `convolution_backward`?https://github.com/pytorch/pytorch/pull/65219

Reply:
automatic reduction only works if shapes are broadcastable. [3] and [1, 3, 128, 128] aren't broadcastable because we pad 1 from the left. [1, 1, 1, 3] and [1, 3, 128, 128] don't match in 3rd and 4th dim. It works for CUDA because we reshape [3] to [1, 3, 1, 1] before we add the output of a conv to it, so we have to manually either reshape or sum. I just did `sum`.

Reply:
> We may just get the last output from the generic convolution_backward

That seems to only compute gradients for input and weight? 
```
- func: cudnn_convolution_backward(Tensor self, Tensor grad_output, Tensor weight, int[] padding, int[] stride, int[] dilation, int groups, bool benchmark, bool deterministic, bool allow_tf32, bool[2] output_mask) -> (Tensor, Tensor)
```

Reply:
in case you want to reproduce the issue outside of TB, this example should do:

```
dev = 'lazy'
inp = torch.rand(1, 3, 128, 128, requires_grad = True).to(device=dev)
in_channels, out_channels = 3, 32
grad = torch.rand(1, 32, 63, 63).to(device=dev)
conv = torch.nn.Conv2d(in_channels, out_channels, kernel_size=3, stride=2, bias=True)
conv.to(dev)
conv_out = conv(inp)
conv_out.backward(grad)
```

Reply:
> automatic reduction only works if shapes are broadcastable. [3] and [1, 3, 128, 128] aren't broadcastable because we pad 1 from the left. [1, 1, 1, 3] and [1, 3, 128, 128] don't match in 3rd and 4th dim. It works for CUDA because we reshape [3] to [1, 3, 1, 1] before we add the output of a conv to it, so we have to manually either reshape or sum. I just did `sum` for testing. Now I can run alexnet_train.

Thanks for the explanation, now I can understand better. When you say use `sum`, I guess the implication is `sum` does auto broadcasting, but how does it affect the actual return value?

Reply:
> > We may just get the last output from the generic convolution_backward
> 
> That seems to only compute gradients for input and weight?
> 
> ```
> - func: cudnn_convolution_backward(Tensor self, Tensor grad_output, Tensor weight, int[] padding, int[] stride, int[] dilation, int groups, bool benchmark, bool deterministic, bool allow_tf32, bool[2] output_mask) -> (Tensor, Tensor)
> ```

Yes, that's my understanding too.

Reply:
@desertfire 
> guess the implication is sum does auto broadcasting

It does the check that `fwd_op` input's shape can be broadcasted to output's shape (which is also gradient's shape). Then it actually runs automatic `sum` (which is a bwd operation to broadcast) to reduce the gradient into the shape of the input. In our case, the input shape is [3] and the gradient shape [1, 3, 128, 128]. So `is_expandable` check in
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L682-L689 fails.

Since we can't rely on the automatic reduction (`sum`) we either have to do it manually which I did in this fix. This is frankly the easiest because the fix is very local to the backward function.

I mentioned we could probably try imitating what CUDA does (e.g. a manual reshape of [3] to [1, 3, 1, 1] but I'm not 100% sure how to pull it off, since by the time we lower to TS we already built the backward graph and for convolution_overrideable we would only run `convolution_overrideable_backward` and that takes [3] bias. I don't know how to "insert" `view` that would be visible to the autograd if it makes sense.

Reply:
> Yes, that's my understanding too.

Yes, so AFAIU we need to handle bias ourselves (outside of `cudnn_convolution_backward`). thankfully, the bwd of addition is just the gradient itself. The only mishap here is with shapes. 😢